pkg/controller/miganalytic/volume_adjustment.go

```diff
@@ -0,0 +1,249 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package miganalytic
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DFCommandExecutor defines an executor responsible for running DF
+type DFCommandExecutor interface {
+	Execute(map[string][]MigAnalyticPersistentVolumeDetails) ([]DFOutput, error)
+}
+
+// PersistentVolumeAdjuster defines volume adjustment context
+type PersistentVolumeAdjuster struct {
+	Owner          *migapi.MigAnalytic
+	Client         client.Client
+	DFExecutor     DFCommandExecutor
+	StatusRefCache map[string]*migapi.MigAnalyticNamespace
+}
+
+// VolumeAdjuster conditions
+const (
+	ExtendedPVAnalysisFailed = "ExtendedPVAnalysisFailed"
+)
+
+// VolumeAdjuster reasons
+const (
+	FailedRunningDf = "FailedRunningDf"
+)
+
+// DFBaseUnit defines supported sizes for df command
+type DFBaseUnit string
+
+// Supported base units used for df command
+const (
+	DecimalSIGiga = DFBaseUnit("GB")
+	DecimalSIMega = DFBaseUnit("MB")
+	BinarySIMega  = DFBaseUnit("M")
+	BinarySIGiga  = DFBaseUnit("G")
+)
+
+// DFCommand represents a df command
+type DFCommand struct {
+	// stdout from df
+	StdOut string
+	// stderr from df
+	StdErr string
+	// Base unit used for df
+	BlockSize DFBaseUnit
+	// BaseLocation defines path where volumes can be found
+	BaseLocation string
+}
+
+// DFDiskPath defines format of expected path of the volume present on Pod
+const DFDiskPath = "%s/%s/volumes/*/%s"
+
+// DFOutput defines structured output of df per PV
+type DFOutput struct {
+	Node            string
+	Name            string
+	Namespace       string
+	UsagePercentage int64
+	TotalSize       resource.Quantity
+	IsError         bool
+}
+
+// convertDFQuantityToKubernetesResource converts a quantity present in df output to resource.Quantity
+func (cmd *DFCommand) convertDFQuantityToKubernetesResource(quantity string) (resource.Quantity, error) {
+	var parsedQuantity resource.Quantity
+	unitMatcher, _ := regexp.Compile(
+		fmt.Sprintf(
+			"(\\d+)%s", cmd.BlockSize))
+	matched := unitMatcher.FindStringSubmatch(quantity)
+	if len(matched) != 2 {
+		return parsedQuantity, errors.Errorf("Invalid quantity or block size unit")
+	}
+	switch cmd.BlockSize {
+	case DecimalSIGiga, DecimalSIMega:
+		quantity = strings.ReplaceAll(quantity, "B", "")
+		break
+	case BinarySIGiga, BinarySIMega:
+		quantity = fmt.Sprintf("%si", quantity)
+		break
+	}
+	return resource.ParseQuantity(quantity)
+}
+
+// GetDFOutputForPV given a volume name and pod uid, returns structured df ouput for the volume
+// only works on outputs of commands created by DFCommand.PrepareDFCommand()
+func (cmd *DFCommand) GetDFOutputForPV(volName string, podUID types.UID) (pv DFOutput) {
```

Comment:
I probably would like a unit test for this, I feel like this could be pretty brittle as we are parsing console out

Have you considered using some `awk` to format the output as JSON?

Reply:
@shawn-hurley since we are running DF for all volumes on a node at once, I thought it might be just easier to parse the output in golang. For instance, we will generate command like this:

```sh
df -BM /path-1 /path-2 /path-3 /path-4 /path-5 ...
```

The logic is very simple. We just look for volume names in output lines (StdErr/StdOut).

Reply:
The logic may be simple, but what if we get back something we don't know how to handle? 

I think why I think a unit test is important, is AFAICT we run a bunch of regex's to find the info, let's make sure that if there is a bug in here and someone tries to fix it, that they don't regress on the output that we are expecting.

The reason for JSON or some other format is probably not needed as I doubt DF will be changing its output structure.

Reply:
@shawn-hurley sounds good. Actually there's already a unit test for this function. Looking at it, I realize that I should add more cases to it which handle unparseable inputs.

Reply:
+1 this seems like a great place for unit tests, pretty self contained input/output

Reply:
updated test